File: sdk/csharp/Rin.Client/RinControlClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace Rin.Client;

/// <summary>Thin loopback client for the engine-neutral Control V2 contract.</summary>
public sealed class RinControlClient : IDisposable
{
    public const string ContractVersion = "rin.control/v2";
    public const string DefaultBaseUrl = "http://127.0.0.1:7375";
    public const int MaximumResponseBytes = 8 * 1024 * 1024;

    private const decimal MaxJsonSafeInteger = 9_007_199_254_740_991m;
    private const int MaxJsonDepth = 64;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    };

    private readonly HttpClient httpClient;
    private readonly string baseUrl;
    private readonly string token;
    private readonly TimeSpan timeout;
    private readonly int maxResponseBytes;

    public RinControlClient(RinControlClientOptions options)
        : this(options, CreateHandler())
    {
    }

    internal RinControlClient(
        RinControlClientOptions options,
        HttpMessageHandler handler)
    {
        if (options is null) throw new ArgumentNullException(nameof(options));
        token = ValidateToken(options.Token);
        baseUrl = NormalizeBaseUrl(options.BaseUrl);
        timeout = options.Timeout;
        if (timeout < TimeSpan.FromMilliseconds(50) ||
            timeout > TimeSpan.FromSeconds(120))
        {
            throw new RinConfigurationException(
                "invalid_timeout",
                "Control timeout must be between 50 ms and 120 seconds");
        }
        maxResponseBytes = options.MaxResponseBytes;
        if (maxResponseBytes < 1024 || maxResponseBytes > MaximumResponseBytes)
        {
            throw new RinConfigurationException(
                "invalid_response_limit",
                "Control response limit must be between 1 KiB and 8 MiB");
        }
        httpClient = new HttpClient(
            handler ?? throw new ArgumentNullException(nameof(handler)),
            disposeHandler: true)
        {
            Timeout = Timeout.InfiniteTimeSpan,
        };
        httpClient.DefaultRequestHeaders.Accept.Add(
            new MediaTypeWithQualityHeaderValue("application/json"));
        httpClient.DefaultRequestHeaders.UserAgent.ParseAdd(
            $"rin-control-csharp/{RinClient.ClientVersion}");
    }

    public async Task<JsonElement> InfoAsync(
        CancellationToken cancellationToken = default)
    {
        var info = await RequestAsync(
            HttpMethod.Get,
            "/control/v2/info",
            null,
            cancellationToken).ConfigureAwait(false);
        if (info.ValueKind != JsonValueKind.Object ||
            !info.TryGetProperty("contract_version", out var contract) ||
            contract.ValueKind != JsonValueKind.String ||
            contract.GetString() != ContractVersion)
        {
            throw new RinProtocolException(
                "control_contract_mismatch",
                "Control Daemon returned an unsupported contract");
        }
        return info;
    }

    public Task<JsonElement> ListWorldsAsync(
        CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/worlds", new Dictionary<string, object>(), cancellationToken);

    public Task<JsonElement> ListActorsAsync(
        object input,
        CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/actors", input, cancellationToken);

    public Task<JsonElement> GetActorAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/actor", input, cancellationToken);

    public Task<JsonElement> WaitActorAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/wait-actor", input, cancellationToken);

    public Task<JsonElement> ObserveActorAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/observe", input, cancellationToken);

    public Task<JsonElement> ListCapabilitiesAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/capabilities", input, cancellationToken);

    public Task<JsonElement> DescribeCapabilityAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/capability", input, cancellationToken);

    public Task<JsonElement> AcquireControllerAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/controllers/acquire", input, cancellationToken);

    public Task<JsonElement> RenewControllerAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/controllers/renew", input, cancellationToken);

    public Task<JsonElement> ReleaseControllerAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/controllers/release", input, cancellationToken);

    public Task<JsonElement> GetControllerAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/controllers/get", input, cancellationToken);

    public Task<JsonElement> SubmitActionAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/actions/submit", input, cancellationToken);

    public Task<JsonElement> ConfirmActionAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/actions/confirm", input, cancellationToken);

    public Task<JsonElement> GetOperationAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/operations/get", input, cancellationToken);

    public Task<JsonElement> WaitOperationAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/operations/wait", input, cancellationToken);

    public Task<JsonElement> CancelOperationAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/operations/cancel", input, cancellationToken);

    public Task<JsonElement> SetEmergencyStopAsync(object input, CancellationToken cancellationToken = default) =>
        PostAsync("/control/v2/emergency-stop", input, cancellationToken);

    public void Dispose() => httpClient.Dispose();

    private static HttpMessageHandler CreateHandler() => new HttpClientHandler
    {
        AllowAutoRedirect = false,
        AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
    };

    private Task<JsonElement> PostAsync(
        string path,
        object input,
        CancellationToken cancellationToken)
    {
        if (input is null) throw new ArgumentNullException(nameof(input));
        return RequestAsync(HttpMethod.Post, path, input, cancellationToken);
    }

    private async Task<JsonElement> RequestAsync(
        HttpMethod method,
        string path,
        object? input,
        CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, baseUrl + path);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        if (input is not null) request.Content = JsonContent(input);

        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        deadline.CancelAfter(timeout);
        HttpResponseMessage response;
        try
        {
            response = await httpClient.SendAsync(
                request,
                HttpCompletionOption.ResponseHeadersRead,
                deadline.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException exception)
            when (!cancellationToken.IsCancellationRequested)
        {
            throw new RinTransportException(
                "transport_timeout",
                "Control Daemon request timed out",
                exception);
        }
        catch (HttpRequestException exception)
        {
            throw new RinTransportException(
                "transport_failed",
                "Control Daemon is unavailable",
                exception);
        }

        using (response)
        {
            var status = (int)response.StatusCode;
            if (status is >= 300 and < 400)
            {
                throw new RinTransportException(
                    "redirect_rejected",
                    "Control Daemon attempted to redirect");
            }
            if (!string.Equals(
                response.Content.Headers.ContentType?.MediaType,
                "application/json",
                StringComparison.OrdinalIgnoreCase))
            {
                throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon response must be application/json");
            }
            if (response.Content.Headers.ContentLength is long declared &&
                (declared < 0 || declared > maxResponseBytes))
            {
                throw new RinProtocolException(
                    "response_too_large",
                    "Control Daemon response exceeds the configured limit");
            }

            byte[] raw;
            try
            {
                raw = await ReadBoundedAsync(
                    response.Content,
                    deadline.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException exception)
                when (!cancellationToken.IsCancellationRequested)
            {
                throw new RinTransportException(
                    "transport_timeout",
                    "Control Daemon request timed out",
                    exception);
            }
            catch (Exception exception)
                when (exception is HttpRequestException or IOException)
            {
                throw new RinTransportException(
                    "transport_failed",
                    "Control Daemon response could not be read",
                    exception);
            }
            return DecodeResponse(raw, status);
        }
    }

    private ByteArrayContent JsonContent(object input)
    {
        byte[] encoded;
        try
        {
            encoded = JsonSerializer.SerializeToUtf8Bytes(
                input,
                input.GetType(),
                JsonOptions);
            using var document = JsonDocument.Parse(encoded);
            if (document.RootElement.ValueKind != JsonValueKind.Object)
            {
                throw new RinProtocolException(
                    "invalid_request",
                    "Control payload must be an object");
            }
            ValidateRequestJson(document.RootElement, 0);
        }
        catch (Exception exception)
            when (exception is JsonException or NotSupportedException or ArgumentException)
        {
            throw new RinProtocolException(
                "invalid_request",
                "Control payload is not JSON serializable",
                exception);
        }
        var content = new ByteArrayContent(encoded);
        content.Headers.ContentType =
            new MediaTypeHeaderValue("application/json") { CharSet = "utf-8" };
        return content;
    }

    private async Task<byte[]> ReadBoundedAsync(
        HttpContent content,
        CancellationToken cancellationToken)
    {
#if NETSTANDARD2_0
        using var stream = await content.ReadAsStreamAsync().ConfigureAwait(false);
#else
        await using var stream = await content
            .ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
#endif
        using var output = new MemoryStream();
        var buffer = new byte[8192];
        while (true)
        {
#if NETSTANDARD2_0
            var count = await stream.ReadAsync(
                buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false);
#else
            var count = await stream.ReadAsync(
                buffer.AsMemory(0, buffer.Length), cancellationToken).ConfigureAwait(false);
#endif
            if (count == 0) break;
            if (output.Length + count > maxResponseBytes)
            {
                throw new RinProtocolException(
                    "response_too_large",
                    "Control Daemon response exceeds the configured limit");
            }
            output.Write(buffer, 0, count);
        }
        return output.ToArray();
    }

    private static JsonElement DecodeResponse(byte[] raw, int status)
    {
        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(raw);
        }
        catch (JsonException exception)
        {
            if (status < 200 || status >= 300)
            {
                throw new RinApiException(
                    ErrorCode(status),
                    "Control Daemon request failed",
                    status);
            }
            throw new RinProtocolException(
                "invalid_response",
                "Control Daemon returned invalid JSON",
                exception);
        }
        using (document)
        {
            var root = document.RootElement;
            if (root.ValueKind is not (JsonValueKind.Object or JsonValueKind.Array))
            {
                throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon response must be an object or array");
            }
            if (status < 200 || status >= 300)
            {
                var code = root.ValueKind == JsonValueKind.Object &&
                    root.TryGetProperty("code", out var codeValue) &&
                    codeValue.ValueKind == JsonValueKind.String
                        ? RinException.SafeText(codeValue.GetString(), 96, ErrorCode(status))
                        : ErrorCode(status);
                var message = root.ValueKind == JsonValueKind.Object &&
                    root.TryGetProperty("error", out var errorValue) &&
                    errorValue.ValueKind == JsonValueKind.String
                        ? RinException.SafeText(
                            errorValue.GetString(),
                            500,
                            "Control Daemon request failed")
                        : "Control Daemon request failed";
                throw new RinApiException(code, message, status);
            }
            return root.Clone();
        }
    }

    private static void ValidateRequestJson(JsonElement value, int depth)
    {
        if (depth > MaxJsonDepth)
        {
            throw new RinProtocolException(
                "invalid_request",
                "Control payload exceeds the JSON nesting limit");
        }
        switch (value.ValueKind)
        {
            case JsonValueKind.Object:
                foreach (var property in value.EnumerateObject())
                {
                    ValidateRequestJson(property.Value, depth + 1);
                }
                return;
            case JsonValueKind.Array:
                foreach (var item in value.EnumerateArray())
                {
                    ValidateRequestJson(item, depth + 1);
                }
                return;
            case JsonValueKind.Number:
                if (value.TryGetDecimal(out var decimalValue))
                {
                    if (decimal.Truncate(decimalValue) == decimalValue &&
                        (decimalValue < -MaxJsonSafeInteger ||
                         decimalValue > MaxJsonSafeInteger))
                    {
                        throw new RinProtocolException(
                            "invalid_request",
                            "Control payload contains an unsafe JSON integer");
                    }
                    return;
                }
                if (!value.TryGetDouble(out var doubleValue) ||
                    double.IsNaN(doubleValue) ||
                    double.IsInfinity(doubleValue))
                {
                    throw new RinProtocolException(
                        "invalid_request",
                        "Control payload contains a non-finite JSON number");
                }
                if (Math.Truncate(doubleValue) == doubleValue &&
                    (doubleValue < -(double)MaxJsonSafeInteger ||
                     doubleValue > (double)MaxJsonSafeInteger))
                {
                    throw new RinProtocolException(
                        "invalid_request",
                        "Control payload contains an unsafe JSON integer");
                }
                return;
            case JsonValueKind.String:
            case JsonValueKind.True:
            case JsonValueKind.False:
            case JsonValueKind.Null:
                return;
            default:
                throw new RinProtocolException(
                    "invalid_request",
                    "Control payload contains a non-JSON value");
        }
    }

    private static string NormalizeBaseUrl(string? value)
    {
        var raw = (value ?? DefaultBaseUrl).Trim().TrimEnd('/');
        if (!Uri.TryCreate(raw, UriKind.Absolute, out var uri) ||
            uri.Scheme != Uri.UriSchemeHttp ||
            uri.Host.Length == 0 ||
            uri.UserInfo.Length > 0 ||
            uri.Query.Length > 0 ||
            uri.Fragment.Length > 0 ||
            (uri.AbsolutePath.Length > 0 && uri.AbsolutePath != "/") ||
            !HasExplicitPort(uri) ||
            !IsLoopback(uri.Host))
        {
            throw new RinConfigurationException(
                "invalid_base_url",
                "Control Daemon URL must be a plain loopback HTTP origin with an explicit port");
        }
        return uri.GetLeftPart(UriPartial.Authority);
    }

    private static bool HasExplicitPort(Uri uri)
    {
        var authority = uri.Authority;
        return authority.StartsWith("[", StringComparison.Ordinal)
            ? authority.IndexOf("]:", StringComparison.Ordinal) >= 0
            : authority.LastIndexOf(':') >= 0;
    }

    private static bool IsLoopback(string host) =>
        host.Equals("localhost", StringComparison.OrdinalIgnoreCase) ||
        (IPAddress.TryParse(host, out var address) && IPAddress.IsLoopback(address));

    private static string ValidateToken(string? value)
    {
        var candidate = value ?? string.Empty;
        if (candidate.Length > 4096 ||
            candidate != candidate.Trim() ||
            candidate.IndexOfAny(new[] { '\0', '\r', '\n' }) >= 0 ||
            Encoding.UTF8.GetByteCount(candidate) < 32)
        {
            throw new RinConfigurationException(
                "invalid_token",
                "Control token must be a bounded single-line value containing at least 32 bytes");
        }
        return candidate;
    }

    private static string ErrorCode(int status) => status switch
    {
        400 => "invalid",
        401 or 403 => "forbidden",
        404 => "not_found",
        409 => "conflict",
        410 => "unavailable",
        429 => "capacity",
        _ => "unavailable",
    };
}

File: sdk/java/src/main/java/io/github/sunrioa/rin/RinControlClient.java
package io.github.sunrioa.rin;

import java.io.ByteArrayOutputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpTimeoutException;
import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Flow;

/** Thin loopback client for the engine-neutral Control V2 contract. */
public final class RinControlClient {
    public static final String VERSION = "0.7.0";
    public static final String CONTRACT_VERSION = "rin.control/v2";
    public static final String DEFAULT_BASE_URL = "http://127.0.0.1:7375";
    public static final int MAX_RESPONSE_BYTES = 8 * 1024 * 1024;

    private final String baseUrl;
    private final String token;
    private final Duration timeout;
    private final int maxResponseBytes;
    private final JsonValueCodec codec;
    private final HttpClient http;

    public RinControlClient(String token, JsonValueCodec codec) {
        this(DEFAULT_BASE_URL, token, Duration.ofSeconds(30), MAX_RESPONSE_BYTES, codec);
    }

    public RinControlClient(
            String baseUrl,
            String token,
            Duration timeout,
            int maxResponseBytes,
            JsonValueCodec codec) {
        this.baseUrl = normalizeBaseUrl(baseUrl);
        this.token = validateToken(token);
        this.timeout = Objects.requireNonNull(timeout, "timeout");
        if (timeout.compareTo(Duration.ofMillis(50)) < 0 ||
                timeout.compareTo(Duration.ofSeconds(120)) > 0) {
            throw new RinConfigurationException(
                    "invalid_timeout",
                    "Control timeout must be between 50 ms and 120 seconds");
        }
        if (maxResponseBytes < 1024 || maxResponseBytes > MAX_RESPONSE_BYTES) {
            throw new RinConfigurationException(
                    "invalid_response_limit",
                    "Control response limit must be between 1 KiB and 8 MiB");
        }
        this.maxResponseBytes = maxResponseBytes;
        this.codec = Objects.requireNonNull(codec, "codec");
        this.http = HttpClient.newBuilder()
                .connectTimeout(timeout)
                .followRedirects(HttpClient.Redirect.NEVER)
                .build();
    }

    public CompletableFuture<Map<String, Object>> info() {
        return requestObject("GET", "/control/v2/info", null).thenApply(info -> {
            if (!CONTRACT_VERSION.equals(info.get("contract_version"))) {
                throw new RinProtocolException(
                        "control_contract_mismatch",
                        "Control Daemon returned an unsupported contract");
            }
            return info;
        });
    }

    public CompletableFuture<List<Map<String, Object>>> listWorlds() {
        return requestObjectList("/control/v2/worlds", Map.of());
    }

    public CompletableFuture<List<Map<String, Object>>> listActors(Map<String, ?> input) {
        return requestObjectList("/control/v2/actors", input);
    }

    public CompletableFuture<Map<String, Object>> getActor(Map<String, ?> input) {
        return postObject("/control/v2/actor", input);
    }

    public CompletableFuture<Map<String, Object>> waitActor(Map<String, ?> input) {
        return postObject("/control/v2/wait-actor", input);
    }

    public CompletableFuture<Map<String, Object>> observeActor(Map<String, ?> input) {
        return postObject("/control/v2/observe", input);
    }

    public CompletableFuture<Map<String, Object>> listCapabilities(Map<String, ?> input) {
        return postObject("/control/v2/capabilities", input);
    }

    public CompletableFuture<Map<String, Object>> describeCapability(Map<String, ?> input) {
        return postObject("/control/v2/capability", input);
    }

    public CompletableFuture<Map<String, Object>> acquireController(Map<String, ?> input) {
        return postObject("/control/v2/controllers/acquire", input);
    }

    public CompletableFuture<Map<String, Object>> renewController(Map<String, ?> input) {
        return postObject("/control/v2/controllers/renew", input);
    }

    public CompletableFuture<Map<String, Object>> releaseController(Map<String, ?> input) {
        return postObject("/control/v2/controllers/release", input);
    }

    public CompletableFuture<Map<String, Object>> getController(Map<String, ?> input) {
        return postObject("/control/v2/controllers/get", input);
    }

    public CompletableFuture<Map<String, Object>> submitAction(Map<String, ?> input) {
        return postObject("/control/v2/actions/submit", input);
    }

    public CompletableFuture<Map<String, Object>> confirmAction(Map<String, ?> input) {
        return postObject("/control/v2/actions/confirm", input);
    }

    public CompletableFuture<Map<String, Object>> getOperation(Map<String, ?> input) {
        return postObject("/control/v2/operations/get", input);
    }

    public CompletableFuture<Map<String, Object>> waitOperation(Map<String, ?> input) {
        return postObject("/control/v2/operations/wait", input);
    }

    public CompletableFuture<Map<String, Object>> getTaskTimeline(Map<String, ?> input) {
        return postObject("/control/v2/tasks/timeline/get", input);
    }

    public CompletableFuture<Map<String, Object>> waitTaskTimeline(Map<String, ?> input) {
        return postObject("/control/v2/tasks/timeline/wait", input);
    }

    public CompletableFuture<Map<String, Object>> cancelOperation(Map<String, ?> input) {
        return postObject("/control/v2/operations/cancel", input);
    }

    public CompletableFuture<Map<String, Object>> setEmergencyStop(Map<String, ?> input) {
        return postObject("/control/v2/emergency-stop", input);
    }

    private CompletableFuture<Map<String, Object>> postObject(String path, Map<String, ?> input) {
        return requestObject("POST", path, Objects.requireNonNull(input, "input"));
    }

    private CompletableFuture<List<Map<String, Object>>> requestObjectList(
            String path,
            Map<String, ?> input) {
        return request("POST", path, Objects.requireNonNull(input, "input"))
                .thenApply(RinControlClient::asObjectList);
    }

    CompletableFuture<Map<String, Object>> requestObject(
            String method,
            String path,
            Map<String, ?> input) {
        return request(method, path, input).thenApply(RinControlClient::asObject);
    }

    private CompletableFuture<Object> request(
            String method,
            String path,
            Map<String, ?> input) {
        HttpRequest.BodyPublisher body = HttpRequest.BodyPublishers.noBody();
        HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .timeout(timeout)
                .header("Accept", "application/json")
                .header("Authorization", "Bearer " + token)
                .header("User-Agent", "rin-control-java/" + VERSION);
        if (input != null) {
            JsonValues.validateRequest(input);
            final String encoded;
            try {
                encoded = codec.encodeObject(input);
            } catch (Exception exception) {
                throw new RinProtocolException(
                        "invalid_request",
                        "Control payload is not JSON serializable",
                        exception);
            }
            if (encoded == null || hasUnpairedSurrogate(encoded)) {
                throw new RinProtocolException(
                        "invalid_request",
                        "Control JSON codec returned invalid output");
            }
            body = HttpRequest.BodyPublishers.ofString(encoded, StandardCharsets.UTF_8);
            builder.header("Content-Type", "application/json; charset=utf-8");
        }
        builder.method(method, body);

        CompletableFuture<HttpResponse<byte[]>> network = http.sendAsync(
                builder.build(),
                ignored -> new BoundedBodySubscriber(maxResponseBytes));
        CompletableFuture<Object> result = new CompletableFuture<>();
        network.whenComplete((response, failure) -> {
            if (failure != null) {
                Throwable cause = unwrap(failure);
                if (cause instanceof RinException) {
                    result.completeExceptionally(cause);
                } else if (cause instanceof HttpTimeoutException) {
                    result.completeExceptionally(new RinTransportException(
                            "transport_timeout",
                            "Control Daemon request timed out",
                            cause));
                } else {
                    result.completeExceptionally(new RinTransportException(
                            "transport_failed",
                            "Control Daemon is unavailable",
                            cause));
                }
                return;
            }
            try {
                result.complete(decodeResponse(response));
            } catch (RuntimeException exception) {
                result.completeExceptionally(exception);
            }
        });
        result.whenComplete((ignored, failure) -> {
            if (result.isCancelled()) network.cancel(true);
        });
        return result;
    }

    private Object decodeResponse(HttpResponse<byte[]> response) {
        int status = response.statusCode();
        if (status >= 300 && status < 400) {
            throw new RinTransportException(
                    "redirect_rejected",
                    "Control Daemon attempted to redirect");
        }
        String mediaType = response.headers()
                .firstValue("Content-Type")
                .orElse("")
                .split(";", 2)[0]
                .strip()
                .toLowerCase(Locale.ROOT);
        if (!"application/json".equals(mediaType)) {
            throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon response must be application/json");
        }
        String contentLength = response.headers().firstValue("Content-Length").orElse(null);
        if (contentLength != null) {
            final long declared;
            try {
                declared = Long.parseLong(contentLength);
            } catch (NumberFormatException exception) {
                throw new RinProtocolException(
                        "invalid_response",
                        "Control Daemon returned an invalid Content-Length",
                        exception);
            }
            if (declared < 0 || declared > maxResponseBytes) {
                throw new RinProtocolException(
                        "response_too_large",
                        "Control Daemon response exceeds the configured limit");
            }
        }

        final String json;
        try {
            json = StandardCharsets.UTF_8.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                    .decode(ByteBuffer.wrap(response.body()))
                    .toString();
        } catch (CharacterCodingException exception) {
            throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon returned invalid UTF-8",
                    exception);
        }

        final Object value;
        try {
            value = codec.decodeValue(json);
        } catch (Exception exception) {
            if (status < 200 || status >= 300) {
                throw new RinApiException(
                        errorCode(status),
                        "Control Daemon request failed",
                        status,
                        "");
            }
            throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon returned invalid JSON",
                    exception);
        }
        if (!(value instanceof Map<?, ?>) && !(value instanceof List<?>)) {
            throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon response must be an object or array");
        }
        if (status < 200 || status >= 300) {
            Map<?, ?> detail = value instanceof Map<?, ?> map ? map : Map.of();
            throw new RinApiException(
                    RinException.safeText(detail.get("code"), 96, errorCode(status)),
                    RinException.safeText(
                            detail.get("error"),
                            500,
                            "Control Daemon request failed"),
                    status,
                    "");
        }
        return value;
    }

    private static Map<String, Object> asObject(Object value) {
        if (!(value instanceof Map<?, ?> map)) {
            throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon response must be an object");
        }
        Map<String, Object> copy = new LinkedHashMap<>();
        for (Map.Entry<?, ?> entry : map.entrySet()) {
            if (!(entry.getKey() instanceof String key)) {
                throw new RinProtocolException(
                        "invalid_response",
                        "Control Daemon response contains a non-string key");
            }
            copy.put(key, entry.getValue());
        }
        return Collections.unmodifiableMap(copy);
    }

    private static List<Map<String, Object>> asObjectList(Object value) {
        if (!(value instanceof List<?> list)) {
            throw new RinProtocolException(
                    "invalid_response",
                    "Control Daemon response must be an array");
        }
        List<Map<String, Object>> copy = new ArrayList<>(list.size());
        for (Object item : list) copy.add(asObject(item));
        return Collections.unmodifiableList(copy);
    }

    private static String normalizeBaseUrl(String value) {
        String raw = value == null || value.isBlank() ? DEFAULT_BASE_URL : value.strip();
        while (raw.endsWith("/")) raw = raw.substring(0, raw.length() - 1);
        final URI uri;
        try {
            uri = new URI(raw);
        } catch (URISyntaxException exception) {
            throw new RinConfigurationException(
                    "invalid_base_url",
                    "Control Daemon URL must be a plain loopback HTTP origin",
                    exception);
        }
        String path = uri.getRawPath();
        if (!"http".equalsIgnoreCase(uri.getScheme()) ||
                uri.getHost() == null ||
                uri.getPort() < 1 ||
                uri.getRawUserInfo() != null ||
                uri.getRawQuery() != null ||
                uri.getRawFragment() != null ||
                (path != null && !path.isEmpty() && !"/".equals(path)) ||
                !isLoopback(uri.getHost())) {
            throw new RinConfigurationException(
                    "invalid_base_url",
                    "Control Daemon URL must be a plain loopback HTTP origin with an explicit port");
        }
        return "http://" + uri.getRawAuthority();
    }

    private static String validateToken(String value) {
        String token = value == null ? "" : value;
        if (token.length() > 4096 ||
                !token.equals(token.strip()) ||
                token.indexOf('\0') >= 0 ||
                token.indexOf('\r') >= 0 ||
                token.indexOf('\n') >= 0 ||
                token.getBytes(StandardCharsets.UTF_8).length < 32) {
            throw new RinConfigurationException(
                    "invalid_token",
                    "Control token must be a bounded single-line value containing at least 32 bytes");
        }
        return token;
    }

    private static boolean isLoopback(String value) {
        String host = value.toLowerCase(Locale.ROOT);
        if (host.equals("localhost") || host.equals("::1") || host.equals("0:0:0:0:0:0:0:1")) {
            return true;
        }
        String[] octets = host.split("\\.", -1);
        if (octets.length != 4) return false;
        try {
            if (Integer.parseInt(octets[0]) != 127) return false;
            for (String octet : octets) {
                int number = Integer.parseInt(octet);
                if (number < 0 || number > 255) return false;
            }
            return true;
        } catch (NumberFormatException ignored) {
            return false;
        }
    }

    private static String errorCode(int status) {
        if (status == 400) return "invalid";
        if (status == 401 || status == 403) return "forbidden";
        if (status == 404) return "not_found";
        if (status == 409) return "conflict";
        if (status == 410) return "unavailable";
        if (status == 429) return "capacity";
        return "unavailable";
    }

    private static boolean hasUnpairedSurrogate(String value) {
        for (int index = 0; index < value.length(); index++) {
            char current = value.charAt(index);
            if (Character.isHighSurrogate(current)) {
                if (index + 1 >= value.length() ||
                        !Character.isLowSurrogate(value.charAt(index + 1))) return true;
                index++;
            } else if (Character.isLowSurrogate(current)) {
                return true;
            }
        }
        return false;
    }

    private static Throwable unwrap(Throwable failure) {
        return failure instanceof CompletionException && failure.getCause() != null
                ? failure.getCause()
                : failure;
    }

    private static final class BoundedBodySubscriber
            implements HttpResponse.BodySubscriber<byte[]> {
        private final int maximum;
        private final ByteArrayOutputStream output;
        private final CompletableFuture<byte[]> body = new CompletableFuture<>();
        private Flow.Subscription subscription;
        private boolean done;

        private BoundedBodySubscriber(int maximum) {
            this.maximum = maximum;
            this.output = new ByteArrayOutputStream(Math.min(maximum, 8192));
        }

        @Override
        public CompletionStage<byte[]> getBody() {
            return body;
        }

        @Override
        public void onSubscribe(Flow.Subscription value) {
            if (subscription != null) {
                value.cancel();
                return;
            }
            subscription = Objects.requireNonNull(value, "subscription");
            subscription.request(1);
        }

        @Override
        public void onNext(List<ByteBuffer> buffers) {
            if (done) return;
            for (ByteBuffer buffer : buffers) {
                int count = buffer.remaining();
                if ((long) output.size() + count > maximum) {
                    fail(new RinProtocolException(
                            "response_too_large",
                            "Control Daemon response exceeds the configured limit"));
                    return;
                }
                byte[] chunk = new byte[count];
                buffer.get(chunk);
                output.write(chunk, 0, count);
            }
            subscription.request(1);
        }

        @Override
        public void onError(Throwable failure) {
            if (done) return;
            done = true;
            body.completeExceptionally(failure);
        }

        @Override
        public void onComplete() {
            if (done) return;
            done = true;
            body.complete(output.toByteArray());
        }

        private void fail(RuntimeException failure) {
            if (done) return;
            done = true;
            subscription.cancel();
            body.completeExceptionally(failure);
        }
    }
}
